src/card.rs

```diff
@@ -0,0 +1,574 @@
+            "8" => Ok(Rank::Num8),
+            "9" => Ok(Rank::Num9),
+            "10" => Ok(Rank::Num10),
+            "jack" => Ok(Rank::Jack),
+            "queen" => Ok(Rank::Queen),
+            "king" => Ok(Rank::King),
+            "ace" => Ok(Rank::Ace),
+            _ => Err(RankError::InvalidName(s.into())),
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct UnorderedCards {
+    cards: BTreeSet<Card>,
+}
+
+impl UnorderedCards {
+    pub fn new() -> Self {
+        default()
+    }
+}
+
+impl CardContainer for UnorderedCards {
+    fn add(&mut self, card: Card) {
+        self.cards.insert(card);
+    }
+
+    fn remove(&mut self, card: &Card) -> bool {
+        self.cards.remove(card)
+    }
+
+    fn contains(&self, card: &Card) -> bool {
+        self.cards.contains(card)
+    }
+
+    fn len(&self) -> u8 {
+        self.cards.len() as u8
+    }
+
+    fn pop_random<R: Rng>(&mut self, rng: &mut R) -> Option<Card> {
+        if self.len() > 0 {
+            let selected = *self.cards.iter().choose(rng).unwrap();
+            if self.remove(&selected) {
+                Some(selected)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+impl FromIterator<Card> for UnorderedCards {
+    fn from_iter<T: IntoIterator<Item = Card>>(iter: T) -> Self {
+        Self {
+            cards: BTreeSet::from_iter(iter),
+        }
+    }
+}
+
+impl IntoIterator for UnorderedCards {
+    type Item = Card;
+    type IntoIter = std::collections::btree_set::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.cards.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a UnorderedCards {
+    type Item = &'a Card;
+    type IntoIter = std::collections::btree_set::Iter<'a, Card>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.cards.iter()
+    }
+}
+
+impl Extend<Card> for UnorderedCards {
+    fn extend<T: IntoIterator<Item = Card>>(&mut self, iter: T) {
+        for card in iter {
+            self.add(card);
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OrderedCards {
+    cards: Vec<Card>,
+}
+
+impl OrderedCards {
+    pub fn new() -> Self {
+        default()
+    }
+}
+
+impl CardContainer for OrderedCards {
+    fn add(&mut self, card: Card) {
+        if !self.cards.contains(&card) {
+            self.cards.push(card);
+        }
+    }
+
+    fn remove(&mut self, card: &Card) -> bool {
+        match self.cards.iter().position(|x| x == card) {
+            Some(pos) => {
+                self.cards.remove(pos);
+                true
+            }
+            _ => false,
+        }
+    }
+
+    fn contains(&self, card: &Card) -> bool {
+        self.cards.contains(card)
+    }
+
+    fn len(&self) -> u8 {
+        self.cards.len() as u8
+    }
+
+    fn pop_random<R: Rng>(&mut self, rng: &mut R) -> Option<Card> {
```

Comment:
Why do we need a random here?

Reply:
Technically this method is not required, but I added it. It will probably not be called.
We could use the thread local random instead?

Reply:
I would suspect we only need random functions on the server-side, right?

Reply:
yeah true. i can remove that.